File: packages/kumoriya_resolver_anime_nexus/lib/src/models/nexus_browser_session.dart
import 'dart:math';

final class NexusBrowserSession {
  NexusBrowserSession._({required this.fingerprint, this.cookieHeader});

  /// Creates a session with an explicit fingerprint (for testing / replay).
  factory NexusBrowserSession.withValues({
    required String fingerprint,
    String? cookieHeader,
  }) =>
      NexusBrowserSession._(
        fingerprint: fingerprint,
        cookieHeader: cookieHeader,
      );

  final String fingerprint;

  final String? cookieHeader;

  NexusBrowserSession withCookieHeader(String? value) {
    return NexusBrowserSession._(
      fingerprint: fingerprint,
      cookieHeader: _mergeCookieHeaders(cookieHeader, value),
    );
  }

  factory NexusBrowserSession.generate() {
    final rng = Random.secure();

    String hex(int bytes) => List<String>.generate(
      bytes,
      (_) => rng.nextInt(256).toRadixString(16).padLeft(2, '0'),
    ).join();

    final fingerprint = _uuidV4(rng, hex);

    // Seed with a synthetic sid cookie – the JS reference
    // (`createBrowserSession`) always starts with `sid=<random_hex>`,
    // and the server binds session state to this value from the very
    // first page scrape onward.  Without it the subsequent WebSocket
    // auth handshake fails with "Authentication failed".
    return NexusBrowserSession._(
      fingerprint: fingerprint,
      cookieHeader: 'sid=${hex(16)}',
    );
  }

  static String _uuidV4(Random rng, String Function(int bytes) hex) {
    final p1 = hex(4); // 8 hex chars
    final p2 = hex(2); // 4 hex chars
    // hex(2) = 4 hex chars; drop the first to keep 3, then prepend '4' → 4xxx
    final p3 = '4${hex(2).substring(1)}';
    final p4a = (8 + rng.nextInt(4)).toRadixString(16);
    // hex(2) = 4 hex chars; drop the first to keep 3 → y + xxx = 4 chars
    final p4b = hex(2).substring(1);
    final p5 = hex(6); // 12 hex chars
    return '$p1-$p2-$p3-$p4a$p4b-$p5';
  }

  static String? _mergeCookieHeaders(String? existing, String? incoming) {
    final merged = <String, String>{};

    void absorb(String? header) {
      if (header == null || header.trim().isEmpty) {
        return;
      }

      for (final part in header.split(';')) {
        final cookie = part.trim();
        if (cookie.isEmpty) {
          continue;
        }
        final separator = cookie.indexOf('=');
        if (separator <= 0) {
          continue;
        }
        merged[cookie.substring(0, separator)] = cookie.substring(
          separator + 1,
        );
      }
    }

    absorb(existing);
    absorb(incoming);

    if (merged.isEmpty) {
      return null;
    }

    return merged.entries
        .map((entry) => '${entry.key}=${entry.value}')
        .join('; ');
  }
}

File: packages/kumoriya_exoplayer/android/src/main/kotlin/dev/kumoriya/exoplayer/nexus/NexusBrowserSession.kt
package dev.kumoriya.exoplayer.nexus

import java.security.SecureRandom

/**
 * Mirror of `NexusBrowserSession.generate()` from the Dart resolver.
 *
 * Every playback session opens a brand new browser-like identity: a UUIDv4
 * fingerprint and a synthetic `sid=<hex>` seed cookie. The real anime.nexus
 * reference implementation (`createBrowserSession` in the web SPA) does the
 * same — the server binds session state to the seeded `sid` from the very
 * first HTTP hit, without it the WebSocket auth handshake fails with
 * `Authentication failed`.
 */
internal class NexusBrowserSession private constructor(
    val fingerprint: String,
    val seedCookieHeader: String,
) {
    companion object {
        fun generate(): NexusBrowserSession {
            val rng = SecureRandom()
            val fingerprint = uuidV4(rng)
            val sid = hex(rng, 16)
            return NexusBrowserSession(
                fingerprint = fingerprint,
                seedCookieHeader = "sid=$sid",
            )
        }

        fun withFingerprint(fingerprint: String, seedCookieHeader: String = ""):
            NexusBrowserSession =
            NexusBrowserSession(fingerprint, seedCookieHeader)

        private fun hex(rng: SecureRandom, bytes: Int): String {
            val buf = ByteArray(bytes)
            rng.nextBytes(buf)
            val sb = StringBuilder(bytes * 2)
            for (b in buf) {
                val v = b.toInt() and 0xFF
                if (v < 0x10) sb.append('0')
                sb.append(v.toString(16))
            }
            return sb.toString()
        }

        /**
         * UUIDv4 faithful to the Dart implementation (randomness concentrated
         * in the same positions, version nibble = `4`, variant nibble in
         * `8..b`). Using [java.util.UUID.randomUUID] would be equivalent for
         * anime.nexus, but we keep the byte layout identical to stay wire
         * compatible when debugging against the Dart resolver.
         */
        private fun uuidV4(rng: SecureRandom): String {
            val p1 = hex(rng, 4) // 8 hex chars
            val p2 = hex(rng, 2) // 4 hex chars
            val p3 = "4" + hex(rng, 2).substring(1) // 4xxx
            val p4a = (8 + rng.nextInt(4)).toString(16) // 8..b
            val p4b = hex(rng, 2).substring(1) // 3 chars
            val p5 = hex(rng, 6) // 12 hex chars
            return "$p1-$p2-$p3-$p4a$p4b-$p5"
        }
    }
}
